MINOR: Add dual write offset metric

hni61223 opened the pull request:
- Add dual write offset metric for ZK migration
- Tested in jconsole

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hni61223 committed:
Fix nits

mumrah reviewed (changes requested):
@hni61223 thanks for the patch! Looking back at the KIP, we specified a lag metric rather than the offset itself. When we're monitoring a migration, we actually care about how far behind the KRaft -> ZK dual write is, so if we report the offset we'll need some calculation in our metrics system.

Let's change this metric name to "ZooKeeperWriteBehindLag" and compute the lag in the gauge function. We'd be comparing it with the lastCommittedRecordOffset

hni61223 committed:
Update metric to ZOO_KEEPER_WRITE_BEHIND_LAG
- update metric to ZOO_KEEPER_WRITE_BEHIND_LAG
- change the metric to be visible only when dual write mode is on

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java:
```diff
@@ -44,6 +45,8 @@ public class QuorumControllerMetrics implements AutoCloseable {
         "ControllerEventManager", "EventQueueTimeMs");
     private final static MetricName EVENT_QUEUE_PROCESSING_TIME_MS = getMetricName(
         "ControllerEventManager", "EventQueueProcessingTimeMs");
+    private final static MetricName ZOO_KEEPER_WRITE_BEHIND_LAG = getMetricName(
```
nit: `ZOOKEEPER` or just `ZK`

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java:
```diff
@@ -109,6 +114,15 @@ public Long value() {
             }
         }));
+
+        if (dualWriteModeState.get()) {
```
This isn't quite right. As written, this atomic boolean won't get updated until we start dual-writing which comes much later than this class being constructed. We'll need to pass in a boolean from ControllerServer indicating if zk migrations are enabled.

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java:
```diff
@@ -109,6 +114,15 @@ public Long value() {
+            registry.ifPresent(r -> r.newGauge(ZOO_KEEPER_WRITE_BEHIND_LAG, new Gauge<Long>() {
+                @Override
+                public Long value() {
+                    return lastCommittedRecordOffset() - dualWriteOffset();
```
Since we won't start updating dualWriteOffset until the migration is done, this returned value will just be equal to the lastCommittedRecordOffset which is ever-increasing. 

I think if dualWriteOffset is zero, we should short-circuit the lag calculation and just return zero. WDYT?

mumrah reviewed (commented):
Thanks for the updates! I left some comments about the behavior of this metric prior to the migration starting.

hni61223 committed:
Fixed turn on metric condition and add test

Fix some nit

mumrah commented on metadata/src/main/java/org/apache/kafka/controller/metrics/QuorumControllerMetrics.java:
```diff
@@ -109,6 +113,18 @@ public Long value() {
+                @Override
+                public Long value() {
+                    // not in dual-write mode: set metric value to 0
+                    if (dualWriteOffset() == 0) return 0L;
+                    // in dual write mode
+                    return lastCommittedRecordOffset() - dualWriteOffset();
```
style nit: let's just do an if and else here rather than the fall-through

mumrah commented on metadata/src/test/java/org/apache/kafka/controller/metrics/QuorumControllerMetricsTest.java:
```diff
@@ -57,11 +57,37 @@ public void testMetricNames() {
+        try {
+            try (QuorumControllerMetrics metrics = new QuorumControllerMetrics(Optional.of(registry), time, true)) {
+                ControllerMetricsTestUtils.assertMetricsForTypeEqual(registry, "kafka.controller",
+                        new HashSet<>(Arrays.asList(
+                                "kafka.controller:type=ControllerEventManager,name=EventQueueTimeMs",
+                                "kafka.controller:type=ControllerEventManager,name=EventQueueProcessingTimeMs",
+                                "kafka.controller:type=KafkaController,name=ActiveControllerCount",
+                                "kafka.controller:type=KafkaController,name=LastCommittedRecordOffset",
+                                "kafka.controller:type=KafkaController,name=LastAppliedRecordLagMs",
+                                "kafka.controller:type=KafkaController,name=LastAppliedRecordOffset",
+                                "kafka.controller:type=KafkaController,name=LastAppliedRecordTimestamp",
+                                "kafka.controller:type=KafkaController,name=LastAppliedRecordOffset",
+                                "kafka.controller:type=KafkaController,name=ZKWriteBehindLag"
+                        )));
+            }
```
nit: too much whitespace. each indentation level should be 4 spaces (no tabs)

mumrah commented on metadata/src/test/java/org/apache/kafka/metadata/migration/KRaftMigrationDriverTest.java:
```diff
@@ -214,8 +233,9 @@ public void testOnlySendNeededRPCsToBrokers() throws Exception {
             metadataPublisher -> { },
             new MockFaultHandler("test"),
             quorumFeatures,
+            metrics,
             mockTime
-        )) {
+            )) {
```
nit: looks like your IDE maybe reformatted some things. can you revert the whitespace and formatting changes?

mumrah commented on metadata/src/test/java/org/apache/kafka/metadata/migration/KRaftMigrationDriverTest.java:
```diff
@@ -344,8 +365,8 @@ public void testShouldNotMoveToNextStateIfControllerNodesAreNotReadyToMigrate()
-            mockTime
-        )) {
+            metrics,
+            mockTime)) {
```
nit: let's keep the style like:

```
foo(

) {

}
```

so in this case, the `)) {` should stay on its own line

mumrah reviewed (commented):
Thanks! this looks great, just a few style nit-picks

mumrah reviewed (approved):
LGTM

mumrah commented:
Failed tests from the build look unrelated

<img width="1144" alt="image" src="https://github.com/apache/kafka/assets/55116/b49741a5-8d59-40b1-b150-0348ec20828e">
